ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -543,6 +582,70 @@ public void onInstanceConfigChange(List<InstanceConfig> configs, NotificationCon
       }
     }
 
+    /**
+     * Triggered whenever the IdealState in current data center has changed (for now, it is usually updated by Helix Bootstrap tool).
+     * @param idealState a list of {@link IdealState} that specifies ideal location of replicas.
+     * @param changeContext the {@link NotificationContext} associated.
+     * @throws InterruptedException
+     */
+    @Override
+    public void onIdealStateChange(List<IdealState> idealState, NotificationContext changeContext)
+        throws InterruptedException {
+      if (!idealStateInitialized.get()) {
+        logger.info("Received initial notification for IdealState change from {}", dcName);
+        idealStateInitialized.set(true);
+      } else {
+        logger.info("IdealState change triggered from {}", dcName);
+      }
+      synchronized (notificationLock) {
+        // rebuild the entire partition-to-resource map in current dc
+        partitionToResourceNameByDc.put(dcName, new ConcurrentHashMap<>());
+        for (IdealState state : idealState) {
+          String resourceName = state.getResourceName();
+          for (String partitionStr : state.getPartitionSet()) {
+            partitionToResourceNameByDc.get(dcName).put(partitionStr, resourceName);
+          }
+        }
+        helixClusterManagerMetrics.idealStateChangeTriggerCount.inc();
+      }
+    }
+
+    /**
+     * Triggered whenever there is a change in the list of live instances.
+     * @param liveInstances the list of all live instances (not a change set) at the time of this call.
+     * @param changeContext the {@link NotificationContext} associated.
+     */
+    @Override
+    public void onLiveInstanceChange(List<LiveInstance> liveInstances, NotificationContext changeContext) {
+      try {
+        logger.trace("Live instance change triggered from {} with: {}", dcName, liveInstances);
```

Comment:
trace seems too low, how often will this happen?

Reply:
I can change it to DEBUG. The reason I don't choose INFO is it contains all live instances (~200) in a single log message which makes it too long. It happens when any server restarts or  shuts down.